go/vt/vttablet/tabletserver/state_manager.go

```diff
@@ -0,0 +1,544 @@
+		case target.Shard != sm.target.Shard:
+			return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid shard %v", target.Shard)
+		case target.TabletType != sm.target.TabletType:
+			for _, otherType := range sm.alsoAllow {
+				if target.TabletType == otherType {
+					return nil
+				}
+			}
+			return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "invalid tablet type: %v, want: %v or %v", target.TabletType, sm.target.TabletType, sm.alsoAllow)
+		}
+	} else {
+		if !tabletenv.IsLocalContext(ctx) {
+			return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "No target")
+		}
+	}
+	return nil
+}
+
+func (sm *stateManager) serveMaster() error {
+	sm.watcher.Close()
+	sm.hr.Close()
+
+	if err := sm.connect(); err != nil {
+		return err
+	}
+
+	sm.hw.Open()
+	sm.tracker.Open()
+	if err := sm.te.AcceptReadWrite(); err != nil {
+		return err
+	}
+	sm.messager.Open()
+	sm.setState(topodatapb.TabletType_MASTER, StateServing)
+	return nil
+}
+
+func (sm *stateManager) unserveMaster() error {
+	sm.unserveCommon()
+
+	sm.watcher.Close()
+	sm.hr.Close()
+
+	if err := sm.connect(); err != nil {
+		return err
+	}
+
+	sm.hw.Open()
+	sm.tracker.Open()
+	sm.setState(topodatapb.TabletType_MASTER, StateNotServing)
+	return nil
+}
+
+func (sm *stateManager) serveNonMaster(wantTabletType topodatapb.TabletType) error {
+	sm.messager.Close()
+	sm.tracker.Close()
+	sm.hw.Close()
+	sm.se.MakeNonMaster()
+
+	if err := sm.connect(); err != nil {
+		return err
+	}
+
+	if err := sm.te.AcceptReadOnly(); err != nil {
+		return err
+	}
+	sm.hr.Open()
+	sm.watcher.Open()
+	sm.setState(wantTabletType, StateServing)
+	return nil
+}
+
+func (sm *stateManager) unserveNonMaster(wantTabletType topodatapb.TabletType) error {
+	sm.unserveCommon()
+
+	sm.tracker.Close()
+	sm.hw.Close()
+	sm.se.MakeNonMaster()
+
+	if err := sm.connect(); err != nil {
+		return err
+	}
+
+	sm.hr.Open()
+	sm.watcher.Open()
+	sm.setState(wantTabletType, StateNotServing)
+	return nil
+}
+
+func (sm *stateManager) connect() error {
+	if err := sm.qe.IsMySQLReachable(); err != nil {
+		return err
+	}
+	if err := sm.se.Open(); err != nil {
+		return err
+	}
+	sm.vstreamer.Open()
+	if err := sm.qe.Open(); err != nil {
+		return err
+	}
+	return sm.txThrottler.Open()
+}
+
+func (sm *stateManager) unserveCommon() {
+	sm.messager.Close()
+	sm.te.Close()
+	sm.qe.StopServing()
+	sm.requests.Wait()
+}
+
+func (sm *stateManager) closeAll() {
+	sm.unserveCommon()
+	sm.txThrottler.Close()
+	sm.qe.Close()
+	sm.watcher.Close()
+	sm.tracker.Close()
+	sm.vstreamer.Close()
+	sm.hr.Close()
+	sm.hw.Close()
+	sm.se.Close()
+	sm.setState(topodatapb.TabletType_UNKNOWN, StateNotConnected)
+}
+
+func (sm *stateManager) setTimeBomb() chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		if sm.timebombDuration == 0 {
+			return
+		}
+		tmr := time.NewTimer(sm.timebombDuration)
+		defer tmr.Stop()
+		select {
+		case <-tmr.C:
+			log.Fatal("Shutdown took too long. Crashing")
```

Comment:
test for this?

Reply:
This can't be tested because it will crash the test :).